### PR TITLE
fix(password): make sure min-max input parameters are respected #297

### PR DESCRIPTION
**Pull request by jeandat**
Also add new min/max global configuration parameters for name and password that allow to constrain them within a controlled range and set good defaults.

Also merge default and user provided configurations once and for all in #forRoot() module method. The goal is to avoid doing it in several places when configuration is needed.

### CODE REVIEW

**Comment by jeandat** [comment]
Wouid you mind telling me what's making codecov unhappy? I don't know this tool and don't really understand the report.

File: package.json [hunk 1]
```diff
@@ -63,7 +63,8 @@
     "@angular/forms": "^8.1.2",
     "@angular/material": "^8.1.1",
     "@angular/router": "^8.1.2",
-    "firebase": "^6.3.1"
+    "firebase": "^6.3.1",
+    "lodash": "^4.17.15"
```
Comment (AnthonyNahas) [inline package.json]: why did you move lodash from dev dependencies to normal dependencies ?

File: src/module/components/ngx-auth-firebaseui-user/user.component.ts [hunk 2]
```diff
@@ -45,23 +45,23 @@ export class UserComponent {
...
     const currentUser: User = this.auth.auth.currentUser;
     this.updateFormGroup = new FormGroup({
       name: this.updateNameFormControl = new FormControl(
-        {value: currentUser.displayName, disabled: this.editMode},
+        { value: currentUser.displayName, disabled: this.editMode },
         [
           Validators.required,
-          Validators.minLength(2),
-          Validators.maxLength(25),
+          Validators.minLength(this.config.nameMinLength),
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui-user/user.component.ts]: Did you test that feature ?

I think the config could be here unsync ...

File: src/module/components/ngx-auth-firebaseui-user/user.component.ts [hunk 3]
```diff
@@ -104,18 +104,18 @@ export class UserComponent {
...
-          snackBarMsg.push(`your email has been update to ${user.email}`);
+          snackBarMsg.push(`your email has been updated to ${user.email}`);
         }
 
         if (this.updatePhoneNumberFormControl.dirty) {
           await user.updatePhoneNumber(this.updatePhoneNumberFormControl.value);
           console.log('phone number = ', this.updatePhoneNumberFormControl.value);
-          snackBarMsg.push(`your phone number has been update to ${user.phoneNumber}`);
+          snackBarMsg.push(`your phone number has been updated to ${user.phoneNumber}`);
         }
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui-user/user.component.ts]: thank you 👍

File: src/module/components/ngx-auth-firebaseui/auth.component.html [hunk 4]
```diff
@@ -27,24 +27,17 @@
...
               </mat-error>
               <mat-error *ngIf="sigInPasswordFormControl.hasError('minlength')">
-                The password must be at least 6 characters long.
+                The password must be at least {{ min }} characters long.
               </mat-error>
               <mat-error *ngIf="sigInPasswordFormControl.hasError('maxlength')">
-                The password must not exceed 25 characters
+                The password must not exceed {{ max }} characters
               </mat-error>
             </mat-form-field>
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.html]: many thanks <3

File: src/module/services/auth-process.service.ts [hunk 5]
```diff
@@ -47,13 +46,12 @@ export class AuthProcessService implements ISignInProcess, ISignUpProcess {
...
-              public afa: AngularFireAuth,
-              private _snackBar: MatSnackBar,
-              private _fireStoreService: FirestoreSyncService) {
-    this.config = Object.assign(defaultAuthFirebaseUIConfig, this.config);
-  }
+  constructor(
+    public afa: AngularFireAuth,
+    private _snackBar: MatSnackBar,
+    private _fireStoreService: FirestoreSyncService,
+    @Inject(forwardRef(() => NgxAuthFirebaseUIConfigToken)) public config: NgxAuthFirebaseUIConfig
```
Comment (AnthonyNahas) [inline src/module/services/auth-process.service.ts]: I really appreciate your contribution ❤️

File: src/module/ngx-auth-firebase-u-i.module.ts [hunk 6]
```diff
@@ -125,19 +131,16 @@ export class NgxAuthFirebaseUIModule {
             provide: FirebaseNameOrConfigToken,
             useFactory: appNameFactory
           },
+          { provide: UserProvidedConfigToken, useValue: config },
           {
             provide: NgxAuthFirebaseUIConfigToken,
-            useValue: config
+            useFactory: ngxAuthFirebaseUIConfigFactory,
+            deps: [UserProvidedConfigToken]
```
Comment (AnthonyNahas) [inline src/module/ngx-auth-firebase-u-i.module.ts]: well done 👍

File: package.json (hunk 1 above)
Comment (jeandat) [reply package.json]: Yes, that's not necessary anymore. My bad, I forgot to revert that. I was using it at runtime first but removed it later. it can stay in devDependencies now.

File: src/module/components/ngx-auth-firebaseui-user/user.component.ts (hunk 3 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui-user/user.component.ts]: Your welcome.

File: src/module/components/ngx-auth-firebaseui-user/user.component.ts (hunk 2 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui-user/user.component.ts]: I'm not following, could you elaborate please? Thx.

**Comment by jeandat** [comment]
Do you think you can merge it and release a new version in a timely manner Anthony? I would need that for my project. Thanks a lot.

**Comment by AnthonyNahas** [comment]
sure! Ill ping you here when I am done with the merge

File: src/module/components/ngx-auth-firebaseui-user/user.component.html [hunk 7]
```diff
@@ -50,9 +50,7 @@
                    placeholder="Name"
                    [formControl]="updateNameFormControl">
             <mat-icon matSuffix>person</mat-icon>
-            <mat-hint align="end" aria-live="polite">
-              {{updateNameFormControl.value?.length}} / 25
-            </mat-hint>
+            <mat-hint align="end" aria-live="polite"> {{ updateNameFormControl.value?.length }} / {{ config.nameMaxLength }} </mat-hint>
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui-user/user.component.html]: don't you think that we need some default values... per example for `config.nameMaxLength` .. ?

File: src/module/components/ngx-auth-firebaseui/auth.component.html [hunk 8]
```diff
@@ -130,11 +123,15 @@
...
-                         [formControl]="sigUpPasswordFormControl"
-                         required>
+                  <input
+                    matInput
+                    [type]="toggle.type"
+                    name="password"
+                    [formControl]="sigUpPasswordFormControl"
+                    required
+                    [minlength]="min"
+                    [maxlength]="max"
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.html]: I think since we are using reactive form and we are indicating the min and max length via the form controller.. we can leave this part out

```html

[minlength]="min"
[maxlength]="max"

`
``

File: src/module/components/ngx-auth-firebaseui/auth.component.ts [hunk 9]
```diff
@@ -163,6 +163,12 @@ export class AuthComponent implements OnInit, AfterViewInit, OnChanges, OnDestro
     if (changes.messageOnAuthSuccess || changes.messageOnAuthError) {
       this.updateAuthSnackbarMessages();
     }
+    if (changes.min) {
+      this.min = this.min != null ? Math.max(this.min, this.config.passwordMinLength) : this.config.passwordMinLength;
+    }
+    if (changes.max) {
+      this.max = this.max != null ? Math.min(this.max, this.config.passwordMaxLength) : this.config.passwordMaxLength;
+    }
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.ts]: default values are not working

Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.ts]: why did you choose the `nhOnChanges`'s hook ?
This will be not fired unless you change some inputs of the component.

File: src/module/components/ngx-auth-firebaseui/auth.component.ts [hunk 10]
```diff
@@ -163,6 +163,12 @@ export class AuthComponent implements OnInit, AfterViewInit, OnChanges, OnDestro
     if (changes.messageOnAuthSuccess || changes.messageOnAuthError) {
       this.updateAuthSnackbarMessages();
     }
+    if (changes.min) {
+      this.min = this.min != null ? Math.max(this.min, this.config.passwordMinLength) : this.config.passwordMinLength;
+    }
+    if (changes.max) {
+      this.max = this.max != null ? Math.min(this.max, this.config.passwordMaxLength) : this.config.passwordMaxLength;
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.ts]: I assume that the logic is wrong

`this.min = this.min != null ? Math.max(this.min, this.config.passwordMinLength) : this.config.passwordMinLength;`

in the above part, if `this.min` is not null and `this.min` is bigger than `this.config.passwordMinLength`
then you will pick the wrong value which is the one configured here `this.config.passwordMinLength`

Do you agree with me ?

`this.max = this.max != null ? Math.min(this.max, this.config.passwordMaxLength) : this.config.passwordMaxLength;
`

the same thing here above: 

if `this.max` is smaller than the one in the configuration, then you will pick the wrong value... 


see more below:

```
 max(...values: number[]): number;
    /**
      * Returns the smaller of a set of supplied numeric expressions.
      * @param values Numeric expressions to be evaluated.
      */
    min(...values: number[]): number;
    /**
      * Returns the value of a base expression taken to a specified power.
      * @param x The base value of the expression.
      * @param y The exponent value of the expression.
      */
```

File: src/module/components/ngx-auth-firebaseui-user/user.component.html (hunk 7 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui-user/user.component.html]: There is default ones in `defaultAuthFirebaseUIConfig`

Comment (AnthonyNahas) [reply src/module/components/ngx-auth-firebaseui-user/user.component.html]: you have right! But the default value were not working because min and max have been initialized on changes hook.

I've just changed the hook to on init

File: src/module/components/ngx-auth-firebaseui/auth.component.html (hunk 8 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui/auth.component.html]: Nope, validators will valide the min and max length but not block user input when you have reached max length. You need both. At least for max length.

Comment (AnthonyNahas) [reply src/module/components/ngx-auth-firebaseui/auth.component.html]: YES! Thanks for pointing that! awesome 🍻

File: src/module/components/ngx-auth-firebaseui/auth.component.ts (hunk 9 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui/auth.component.ts]: You are right. We need something in ngOnInit too to initialize min/max if they have not been setted.

File: src/module/components/ngx-auth-firebaseui-user/user.component.html (hunk 7 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui-user/user.component.html]: You need both in fact. ngOnInit for the case where min/max is not provided and ngOnChanges for the case they are provided and might change over time.

File: src/module/components/ngx-auth-firebaseui/auth.component.ts (hunk 10 above)
Comment (jeandat) [reply src/module/components/ngx-auth-firebaseui/auth.component.ts]: I don't think so. 

> The Math.max() function returns the largest of zero or more numbers.

If `this.min` is `not null` and `>` to `this.config.passwordMinLength`, `Math.min` will return `this.min`.

**Comment by AnthonyNahas** [comment]
@jeandat many many thanks for this PR! 👍 

I've just release [v3.3.0](https://www.npmjs.com/package/ngx-auth-firebaseui)

cheers 🍻
